fastavro/_read_py.py

```diff
@@ -326,6 +330,15 @@ def read_record(decoder, writer_schema, reader_schema=None,
             record[field['name']] = read_data(decoder, field['type'], None,
                                               return_record_name)
     else:
+        # Handle case where reader_schema is an union
+        if isinstance(reader_schema, list):
+            for schema in reader_schema:
+                try:
+                    if match_types(writer_schema, schema):
+                        reader_schema = schema
+                except SchemaResolutionError:
+                    # Try to match the next schemas in the union
+                    pass
```

Comment:
Add the raise `for/else` statement like in `match_schemas` function

Reply:
you're right, I added the condition

Reply:
So the `else` will be evaluated every time, you need a `break` after finding the first `match_types()`.

Reply:
indeed. fixed, thanks !